client_integration_test.go

```diff
@@ -1489,6 +1490,59 @@ func TestClientReadFrom(t *testing.T) {
 	}
 }
 
+// A wrongSizeReader reads zeros and has an arbitrary Size.
+type wrongSizeReader struct{ actual, reported int }
+
+func (r *wrongSizeReader) Read(p []byte) (n int, err error) {
+	if len(p) >= r.actual {
+		n, r.actual = r.actual, 0
+		return n, io.EOF
+	}
+	r.actual -= len(p)
+	return len(p), nil
+}
+
+func (r *wrongSizeReader) Size() int { return r.reported }
+
+// Test File.ReadFrom's handling of a Reader's Size:
+// it should be used as a heuristic for determining concurrency only.
+func TestClientReadFromSizeMismatch(t *testing.T) {
+	const (
+		packetSize = 1024
+		filesize   = 4 * packetSize
+	)
+
+	sftp, cmd := testClient(t, READWRITE, NODELAY, MaxPacketChecked(packetSize), UseConcurrentWrites(true))
+	defer cmd.Wait()
+	defer sftp.Close()
+
+	d, err := ioutil.TempDir("", "sftptest-readfrom-size-mismatch")
+	if err != nil {
+		t.Fatal("cannot create temp dir:", err)
+	}
+	defer os.RemoveAll(d)
+
+	for i, reported := range []int{-1, filesize - 100, filesize, filesize + 100} {
+		r := &wrongSizeReader{filesize, reported}
+
+		f := path.Join(d, fmt.Sprintf("writeTest%d", i))
+		w, err := sftp.Create(f)
+		if err != nil {
+			t.Fatal("unexpected error:", err)
+		}
+		defer w.Close()
```

Comment:
These `defer`s only trigger at the end of the function, not the end of each loop. It’s only four handles right now so it’s not so bad, but it’s a good idea to always be wary of any `defer` that appears in a loop.

Reply:
Replaced by subtests. Btw., the test just above hold more files open than this one...